semana18/aula1/FutureBook/src/business/usecases/LoginUseCase.ts

```diff
@@ -0,0 +1,41 @@
+import {UserGettersGateway, UserTokenGateway} from "../gateways/userGateway";
+import {BcryptCompareGateway} from "../gateways/bcryptCompareGateway";
+import {User} from "../entities/User";
+
+export class LoginUseCase {
+  constructor(
+    private userGettersGateway: UserGettersGateway,
+    private bcryptcompareGateway: BcryptCompareGateway,
+    private userTokenGateway: UserTokenGateway
+  ) {}
+  async execute(input: LoginUseCaseInput): Promise<LoginUseCaseOutput> {
+    const user = await this.userGettersGateway.getUserByEmail(input.email);
+
+    if(!user) {
+      throw new Error("Usuário não existe")
+    }
+
+    const isPasswordCorrect = await this.bcryptcompareGateway.compare(
+      input.password,
+      user.getPassword()
+    );
+
+    if(!isPasswordCorrect){
+      throw new Error ("Senha ou email incorretos")
+    }
+    const token = this.userTokenGateway.generateToken({...user});
```

Comment:
Um pequeno comentário: gostei de você ter desconstruído o usuário aqui! Você está seguindo um princípio interessante, chamado de "princípio do bom vizinho", que diz que as classes que se comunicam devem passar os parâmetros da melhor forma possível para quem for receber não precisar fazer muitos tratamentos. Claro que há casos e casos, mas tentar isolar os tratamentos em uma classe só (no caso o use case) é bem inteligente! :D